src/psyclone/psyir/symbols/symbol_table.py

```diff
@@ -993,6 +1083,66 @@ def swap(self, old_symbol, new_symbol):
         self.remove(old_symbol)
         self.add(new_symbol)
 
+    def _remove_routinesymbol(self, symbol):
+        '''
+        Removes the supplied RoutineSymbol from this table when possible.
+        If the symbol is the target of a Call or is a member of an interface
+        definition then removal is only possible if the symbol is shadowing a
+        declaration in an outer scope. In that case the Call or Interface is
+        updated to refer to that instance of the symbol.
```

Comment:
I couldn't remember why I needed this but then I looked at the tests and see that we can't currently flatten symbol tables if they contain separate references to the same imported symbol. (I think this can happen after a call within e.g. a loop has been inlined.) I take your point that `remove` should only "remove" and anything else should be `swap`.

We are therefore using shadowing to make our lives easier when inlining but that then massively trips us up when we come to flatten the symbol tables. As you say elsewhere, the best solution to this may be to ensure that we fully resolve any such symbols by adding the names of the relevant modules to the optimisation script.

Reply:
Somewhere along the way I've improved things such that this change is no longer needed for the integration tests. However, this revealed that the inlining transformation is not correctly handling the case where a symbol in both tables is left unaffected by a merge because it refers to the same global object. What should happen is that any references to such an object be updated so that they all refer to the correct symbol in the current scope. However, that's a problem for InlineTrans rather than SymbolTable.merge so I've added an xfailing test with a TODO for #924.

For now, I've removed the separate handling of 'shadowed' symbols entirely and added a TODO for #2271 to improve the search for anything preventing the removal of a RoutineSymbol. That way, we're no worse than we are without this PR :-)